collect_app/src/main/java/org/odk/collect/android/storage/migration/StorageMigrationService.java

```diff
@@ -0,0 +1,51 @@
+package org.odk.collect.android.storage.migration;
+
+import android.content.Intent;
+
+import androidx.lifecycle.LifecycleService;
+import androidx.localbroadcastmanager.content.LocalBroadcastManager;
+
+import org.jetbrains.annotations.NotNull;
+import org.odk.collect.android.storage.StoragePathProvider;
+import org.odk.collect.android.storage.StorageStateProvider;
+
+public class StorageMigrationService extends LifecycleService {
```

Comment:
I think we should be using an `IntentService` here (https://developer.android.com/guide/components/services#ExtendingIntentService). I've usually used them in the past as they take care of creating a worker threads and also protect us from multiple requests being exectued simultaneously etc. This would potentially mean the service could take care of the lock in `StorageMigrator` (`.isMigrationBeingPerformed`). Also if we do use a plain ol' service I think it needs to call `stopSelf()` when we're done executing - again an `IntentService` just takes care of that for us. 

Sorry if this is research you already did and made the decision for some reason I'm missing!

Reply:
I was thinking about using broadcast receivers vs livedata... could you take a look at the last commit when I replaced broadcast receivers with livedata and share you thought?

Reply:
Using `LiveData` does seem a lot cleaner in the observers (the `Activity` and the `DialogFragment`). I'd want to avoid having the `LiveData` object being static and also avoid having the `StorageMigrator` have to deal with `LiveData` though. This is to avoid us running into problems with tests polluting each other  and also just avoids complicating the `StorageMigrator` with reactive stuff.

What I might do is look at having a seperate `MigrationStatusRepository` like this:

```java
class MigrationStatusRepository {

    private MutableLiveData<StorageMigrationResult> migrationStatus = new MutableLiveData<>();

    public LiveData<StorageMigrationResult> getStatus() {
        return migrationStatus;
    }

    public void setStatus(StorageMigrationResult storageMigrationResult) {
        migrationStatus.setValue(storageMigrationResult);
    }
}
```

The Activity/Fragment would call `getStatus()` to observe and the Service could write the status when `StorageMigrator` completes using `setStatus`. This can be set up as a singleton (Dagger's `@Singleton`) and injected into the Activity, Fragment and Service. That way it can be easily swapped out in tests for any of those objects.

How does that feel? cc/ @lognaturel

Reply:
But the service doesn't know about every single step it just starts migration (one method) sot it might receive the result but not every single status which changes after every step. Looking at your code where you mixed words status and result I think you didn't notice that we have both: one general result and a few statuses to update current progress text.

Reply:
So I would need to 
- inject such an repository to StorageMigrator class
- or divide the migration into a few methods and call all of them in Service one by one (and then deal with statuses in Service)

Reply:
Ah you're absolutely right I missed that somehow. Sorry about that.

Yeah I'd still think about about looking at something lilke the `MigrationStatusRespository`. It could injected into the `StorageMigrator` (probably just as a constructor argument). It could still just have one set/get pair as the "results" could also just be a statuses i.e. the status could change from `MIGRATING_DATABASES` to `SUCCESS` or  from `PREPARING_SCOPED_STORAGE` to `NOT_ENOUGH_SPACE`.

Reply:
Ok I pushed new changes and now I use `LiveData` and a repository class for both results and statuses. I think those statuses is more like my idea so here is a question to @lognaturel because maybe you just want to display one message like `Please wait. it might take a few seconds...` and then we can get rid of it at all otherwise I think we need both.

Reply:
My initial reaction was that I really like the statuses because it gives users a sense of progress. That said, it's not clear that the information is actionable, the migration is going to be really quick for most people and it's a lot of translation effort for a one-time migration. Let's go with "Migration in progress.\nThis may take a few seconds..." I do like the approach you'd come up with for the statuses, though @grzesiek2010. Sorry to see it go.

Reply:
Ok, done.